### PR TITLE
DAO member handling

### PR DESCRIPTION
**Pull request by opny721**
This  PR set the basis for DAO members management and enable task assignment. Still very early but good to close an iteration IMHO

### CODE REVIEW

File: backend/src/dao/controller/member.controller.ts [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+  @DaoAuth(Role.founder, Role.projectManager)
+  list(@Param('daoId') daoId: string, @Body() query: MemberQueryDto): Promise<ExtendedMemberDto[]> {
+    query.daoId = daoId
+    return this.memberService.list(query)
+  }
+
+  @Post('invite')
+  @HttpCode(200)
+  @DaoAuth(Role.founder, Role.projectManager)
+  invite(
```
Comment (ivelin) [inline backend/src/dao/controller/member.controller.ts:28]: @opny721 I need some clarification here. For the project management flow, we have the roles of:
- Project Manager: who creates a ProjectProposal (budget, deadline, KPIs) and submits to DAO for approval
- Contributor: who receives an invite from a Project Manager to take on a task.

Once a task is completed, Project Manager approves and DAO mints sweat tokens to Contributor in exchange for work deliverable.

I am getting confused with the code using terms as Members of the DAO. What do you consider Member of a DAO?

File: backend/src/dao/controller/member.controller.ts [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+    @Param('daoId') daoId: string,
+    @User('userId') userId: string,
+    @Body() invite: MemberInviteDto,
+  ): Promise<MemberDto> {
+    invite.daoId = invite.daoId
```
Comment (ivelin) [inline backend/src/dao/controller/member.controller.ts:33]: Help me unpack the intention of this method.

Someone with Founder or Project Manager role sends an invite to a member to do what?

Is the intention here for a project manager to invite a contributor to join a project or take on a task?

Let's clarify that an invitation to participate in a project does not yet mean that the contributor will receive sweat tokens. That happens only when they produce a deliverable and after project manager approval, they trade it for sweat tokens.

Also a contributor may already have sweat tokens from previous projects they participated in. Or they may be a brand new contributor with no sweat tokens granted yet.

My point is that we need to distinguish between participation in a project and owning sweat tokens.

File: backend/src/dao/controller/member.controller.ts [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+    memberDto.invitedBy = userId
+
+    return this.memberService.update(memberDto)
+  }
+
+  @Delete(':memberId')
+  @HttpCode(200)
+  @UseGuards(RolesGuard)
+  @DaoAuth(Role.founder)
+  delete(@Param('daoId') daoId: string, @Param('memberId') memberId: string): Promise<void> {
```
Comment (ivelin) [inline backend/src/dao/controller/member.controller.ts:80]: Deleting a member from what type of participation? If the intention is to disinvite / kick a contributor out of a project, we should clarify that is the case. 

If the intention is to kick a member out of the DAO by burning their sweat tokens, then that's a different use case.

File: backend/src/dao/controller/project.controller.ts [hunk 4]
```diff
@@ -42,15 +40,19 @@ export class ProjectController {
...
+  ): Promise<ProjectDto> {
     projectDto.projectId = projectId
     return this.projectService.update(projectDto)
   }
 
   @Delete(':projectId')
   @HttpCode(200)
   @Auth(Role.projectManager, Role.founder)
-  delete(@Param('projectId') projectId: string): Promise<void> {
+  delete(@Param('daoId') daoId: string, @Param('projectId') projectId: string): Promise<void> {
```
Comment (ivelin) [inline backend/src/dao/controller/project.controller.ts:55]: How about methods to find all contributors currently participating (after invitation) in a project? We will probably need something in the UI to show project board with task status and contributors assigned to tasks.

File: backend/src/dao/controller/task.controller.ts [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+@ApiBearerAuth()
+@ApiTags('task')
+@Controller('/dao/:daoId/task')
+export class TaskController {
+  constructor(private taskService: TaskService) {}
+
+  @Post('find')
+  @HttpCode(200)
+  @DaoAuth()
+  find(@Param('daoId') daoId: string, @Body() query: TaskQueryDto): Promise<TaskDto[]> {
```
Comment (ivelin) [inline backend/src/dao/controller/task.controller.ts:17]: If I understand this code correctly, tasks are associated with DAOs. Shouldn't tasks be associated with projects instead?

DAO -> Project -> Task

**Review by ivelin — COMMENTED** [review]
Lot's of work in this PR. Dropped a few comments to clarify intention and make sure we are aligned on design.

File: backend/src/dao/controller/member.controller.ts (hunk 1 above)
Comment (opny721) [reply backend/src/dao/controller/member.controller.ts:28]: A member of the DAO is technically a link between a user and a dao, roles such as Founder, PM are attached to a DAO membership and enforced duirng various operations. A member with no explicit role is a contributor to which a PM can assign tasks.

File: backend/src/dao/controller/member.controller.ts (hunk 2 above)
Comment (opny721) [reply backend/src/dao/controller/member.controller.ts:33]: This specific api call just create a member record to track user activities and role inside the DAO, no other meaning so far.
Let's say you own a DAO and invite me as PM. I then lead a project and in turn I invite other users to carry on task. 

So far there are no implicit advantages for users, see it just as a way to assign roles inside a DAO

File: backend/src/dao/controller/member.controller.ts (hunk 3 above)
Comment (opny721) [reply backend/src/dao/controller/member.controller.ts:80]: This is part of generated codebase, let's see if it make sense or not. Probably we will mark a user as inactive or banned, not just remove it

File: backend/src/dao/controller/project.controller.ts (hunk 4 above)
Comment (opny721) [reply backend/src/dao/controller/project.controller.ts:55]: Good suggestion, I will add this to #58 for founders/PMs

File: backend/src/dao/controller/task.controller.ts (hunk 5 above)
Comment (opny721) [reply backend/src/dao/controller/task.controller.ts:17]: Yes the structure is like the above DAO > Project > Task
To simplify a bit the authorization I opted to require explicity the daoId from controller (used to load the members and roles see [@DaoAuth](https://github.com/opny721/sweat-token/blob/feature/39-people-section/backend/src/dao/dao.auth.decorator.ts#L8)). When we have a stable flow, we can improve the API interfaces

File: backend/src/dao/controller/member.controller.ts (hunk 2 above)
Comment (opny721) [reply backend/src/dao/controller/member.controller.ts:33]: > This informal offchain participation probably does not need to be represented in the SporosDAO app in the MPV.

I agree but how do you internally track roles and entity ownership? Eg. User A own a Project, User B is contributing to task X ?

> Handle sweat token projects with multi-week, multi contributor scope. Sweat tokens are non-transferrable.

That's ok, maybe you have different expectations on the implementation approach?

So far I am following a web2-ish approach to provide the base functionalities as per above statement while we will integrate to KaliDAO once the PM/contributor flow is working (I would say task assignment, reporting and task confirmation).

The KaliDAO integration comes later because we store on-chain only the relevant parts (e.g. DAO creation, project creation and assignment, dispute resolution, Sweat Token conversion). 

To recap, my take on this is that the on-chain integration (the legal framework gateway) hooks on a sub-layer that has a few known steps but does not have specific assumptions on how it should work internally. Before jumping into smart contract dev I would see the basic flow validated and improved enough to be reasonably stable. 

What do you think ?

> I imagine we would have a proposal type named SweatEquityProjectProposal with these parameters:

That is a very good hint, if the above points are good for you I would move to (an expanded version of) #4

Comment (ivelin) [reply backend/src/dao/controller/member.controller.ts:33]: Solid!
